### PR TITLE
assorted fixes

### PR DESCRIPTION
**Pull request by vondele**

### CODE REVIEW

**Comment by vondele** [comment]
This appears to be working now, but a critical look is welcome.

local testing with a restart from master and resumed training looks like this: 
```
python easy_train.py --training-dataset=./nodes5000pv2_UHO.binpack --nnue-pytorch=vondele/nnue-pytorch/easy_train --experiment=003 --network-save-period=2 --start-from-stockfish-net=True --resume-training=True
```
and yields:
image
that's far lower Elo than I'm used to, but maybe this is because of the very fast TC (25knodes), which could increase Elo differences. @Sopel97 maybe you could check if this matches your typical results training from master.

[71bda23](https://github.com/Sopel97/nnue-pytorch/pull/2/commits/71bda23e3ae664992f95fa4c1fc5295f95b669c0)

is a workaround for a bug that would need more proper fixing, basically run_games needs more care to ensure the output of the two threads doesn't corrupt each other, this lead to the Elo not updating in the TUI.

**Comment by vondele** [comment]
Running overnight, I seem to find a crash, which I believe is due to Ordo running out of memory. After a few hours 29552220 games are played and a 6GB pgn file is available. However, it appears that ordo uses about 60GB to deal with it. Not 100% sure that this is the reason for this crash report, but out of mem could cause nearly anything:
```
$ cat error.txt 
Traceback (most recent call last):
  File "easy_train.py", line 1829, in <module>
    main()
  File "easy_train.py", line 1807, in main
    Screen.wrapper(app, catch_interrupt=True, arguments=[last_scene, training_runs, network_testing])
  File "/home/vondele/chess/vondele/nnue-pytorch/env/lib/python3.8/site-packages/asciimatics/screen.py", line 1433, in wrapper
    return func(screen, *arguments)
  File "easy_train.py", line 1436, in app
    screen.play(scenes, stop_on_resize=True, start_scene=scene, allow_int=True)
  File "/home/vondele/chess/vondele/nnue-pytorch/env/lib/python3.8/site-packages/asciimatics/screen.py", line 1606, in play
    self.draw_next_frame(repeat=repeat)
  File "/home/vondele/chess/vondele/nnue-pytorch/env/lib/python3.8/site-packages/asciimatics/screen.py", line 1709, in draw_next_frame
    effect.update(self._frame)
  File "easy_train.py", line 1408, in update
    self._update_network_testing_status()
  File "easy_train.py", line 1402, in _update_network_testing_status
    self._network_testing_status.value = self._network_testing.get_status_string()
  File "easy_train.py", line 1018, in get_status_string
    f'Testing run {self._current_test.run_id} epoch {self._current_test.epoch}',
AttributeError: 'NoneType' object has no attribute 'run_id'
^CException ignored in: <module 'threading' from '/usr/lib/python3.8/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 1388, in _shutdown
    lock.acquire()
KeyboardInterrupt: 

```

**Comment by vondele** [comment]
Also `--network-testing-time-per-move` is time per move or per game?

File: scripts/easy_train.py [hunk 1]
```diff
@@ -1433,74 +1446,90 @@ def str2bool(v):
...
-    parser.add_argument("--start-from-model", default=None, type=str, dest='start_from_model', help="Initializes training using the weights from the given .pt model")
+    parser.add_argument("--start-from-model", default=None, type=str, dest='start_from_model', help="Initializes training using the weights from the given .pt .ckpt .nnue model")
     parser.add_argument("--start-from-experiment", default=None, type=str, dest='start_from_experiment', help="Initializes training using the best network from a given experiment (by name). Uses best net from ordo, fallbacks to last.")
-    parser.add_argument("--gpus", type=str, dest='gpus', default='0')
-    parser.add_argument("--runs-per-gpu", type=int, dest='runs_per_gpu', default=1)
-    parser.add_argument("--features", type=str, help="The feature set to use")
+    parser.add_argument("--start-from-stockfish-net", default=False, type=str2bool, dest='start_from_stockfish_net', help="Initializes training using the weights from the .nnue associated with --engine-test-branch")
```
Comment (Sopel97) [inline scripts/easy_train.py]: perhaps `--start-from-engine-test-net` would be a better name

File: scripts/easy_train.py [hunk 2]
```diff
@@ -1433,74 +1446,90 @@ def str2bool(v):
...
+    parser.add_argument("--gpus", type=str, dest='gpus', default='0', help="a GPU ID or a list of GPU IDs.")
+    parser.add_argument("--runs-per-gpu", type=int, dest='runs_per_gpu', default=1, help="To increase the load on strong GPUs run more than one repetition of this experiment and reduce net variance.")
+    parser.add_argument("--features", type=str, default='HalfKAv2_hm^', help="The feature set to use") # TODO can this be made a default based on the nnue-pytorch-branch specified?
```
Comment (Sopel97) [inline scripts/easy_train.py:1475]: yea, possibly by searching for `_default_feature_set_name = 'HalfKAv2_hm^'` in features.py. Needs to happen after nnue-pytorch setup

File: scripts/easy_train.py [hunk 3]
```diff
@@ -1433,74 +1446,90 @@ def str2bool(v):
...
-    parser.add_argument("--build-threads", type=int, default=1, dest='build_threads', help="Number of threads to use for compilation")
+    parser.add_argument("--engine-base-branch", type=str, default='official-stockfish/Stockfish/master', dest='engine_base_branch', help="Path to the commit/branch to use for the engine baseline.")
+    parser.add_argument("--engine-test-branch", type=str, default='official-stockfish/Stockfish/master', dest='engine_test_branch', help="Path to the commit/branch to use for the engine being tested.")
+    parser.add_argument("--nnue-pytorch-branch", type=str, default='glinscott/nnue-pytorch/master', dest='nnue_pytorch_branch', help="Path to the commit/branch to use for the trainer being tested.")
+    parser.add_argument("--build-engine-arch", type=str, default='x86-64-modern', dest='build_engine_arch', help="ARCH to use for engine compilation, e.g. x86-64-avx2 for recent hardware")
+    parser.add_argument("--build-threads", type=int, default=8, dest='build_threads', help="Number of threads to use for compilation")
```
Comment (Sopel97) [inline scripts/easy_train.py]: or something like half of cpu_count?

File: scripts/easy_train.py [hunk 4]
```diff
@@ -1433,74 +1446,90 @@ def str2bool(v):
...
     parser.add_argument("--network-testing-explore-factor", type=float, default=1.5, dest='network_testing_explore_factor', help="Elo error estimates are multiplied by this amount to determine testing candidates.")
-    parser.add_argument("--network-testing-book", type=str, default='./noob_3moves.epd', dest='network_testing_book', help="Path to a suitable book, see https://github.com/official-stockfish/books")
+    parser.add_argument("--network-testing-book", type=str, default='https://github.com/official-stockfish/books/raw/master/UHO_XXL_+0.90_+1.19.epd.zip', dest='network_testing_book', help="Path to a suitable book, or suitable link see https://github.com/official-stockfish/books")
     parser.add_argument("--network-testing-time-per-move", type=float, default=None, dest='network_testing_time_per_move', help="Number of seconds per game")
```
Comment (Sopel97) [inline scripts/easy_train.py]: `--network-testing-time-per-move` should indeed be named `--network-testing-time-per-game`

**Review by Sopel97 — COMMENTED** [review]
Thanks, that's quite a lot of stuff.

Regarding ordo, it does indeed seem to use a lot of memory. Not sure what we can do about this, apart from fixing ordo itself.

That error is a race on self._current_test. I'll need to add some mutexes in places, but will take some time to figure out exactly where. Or alternatively more try except, since we don't really care about the semantics that much. (I get ~-100 elo from scratch after 20 epochs, 25k nodes per move)

Those Elo numbers definitely don't look right. Either something wrong with deserialization, or doesn't actually correctly start from model and instead starts from scratch, or something weird going on.

**Comment by vondele** [comment]
meanwhile I had an idea why the Elo numbers look wrong... is this nnue branch already up to date with SF master SFNNv5? In that case the restart would be quite bad. I'm testing right now.

right, starting from my vondele/Stockfish/lastSFNNv4 branch looks much better, -90Elo for epoch 1. That's in line with expectations.

image

File: scripts/easy_train.py (hunk 4 above)
Comment (vondele) [reply scripts/easy_train.py]: done.

File: scripts/easy_train.py (hunk 2 above)
Comment (vondele) [reply scripts/easy_train.py:1475]: I'll pass on this one, as I don't know this well enough.

**Comment by vondele** [comment]
concerning Ordo... somehow we don't need it. It would be enough to just record the WLD numbers provided by c-chess-cli, since all our games are master vs run_x/epoch_y and the Elo relative to master is easy to compute. That's a bit of a major surgery, but having Ordo as the memory bottleneck of training is not a good idea.

**Comment by Sopel97** [comment]
Alternatively we can add a switch in run_games.py to use a simple calculation instead of ordo (but we should keep the output format compatible)

I'll merge this and see what still needs to be done from there.

**Pull request closed without merging**